### PR TITLE
feat: simple thread pool

### PR DESCRIPTION
**Pull request by olgavrou**
introduced to be used in LAS

will probably evolve to include some sort of thread pool interface

when there is no work to be done threads wait on a conditional variable

if thread pool is initialized with zero size then the tasks are just executed sequentially on the same thread that called the thread pool

### CODE REVIEW

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+#include <queue>
+#include <thread>
+
+// thread pool based off of the simple pool implementation from ccia:
+// https://github.com/anthonywilliams/ccia_code_samples
+
+namespace VW
+{
+template <typename T>
+class thread_safe_queue
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:23]: This is really similar to `queue.h` can we reuse it?

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    std::future<ResultType> result(task.get_future());
+
+    if (_threads.size() == 0)
+    {
+      task();
+      return result;
+    }
+
+    {
+      std::unique_lock<std::mutex> l(_mut);
```
Comment (zwd-ms) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:176]: `push` has lock already?

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+  // Enqueues a func task in the work queue for worker threads to execute.
+  template <typename F, typename... Args>
+  std::future<typename std::result_of<F(Args...)>::type> submit(F&& func, Args&&... args)
+  {
+    using ResultType = typename std::result_of<F(Args...)>::type;
+    std::packaged_task<ResultType()> task(std::bind(std::forward<F>(func), std::forward<Args>(args)...));
+    std::future<ResultType> result(task.get_future());
+
+    if (_threads.size() == 0)
```
Comment (zwd-ms) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:169]: what's the use case for initializing a thread pool with zero threads?

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    function_wrapper(function_wrapper&& other) : _impl{std::move(other._impl)} {}
+
+    function_wrapper(const function_wrapper&) = delete;
+
+    function_wrapper(function_wrapper&) = delete;
+
+    function_wrapper& operator=(const function_wrapper&) = delete;
+
+    void operator()() { _impl->call(); }
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:93]: It looks like this means the thread pool cant be used for operations which return values?

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    {
+    }
+
+    function_wrapper() = default;
+
+    function_wrapper(function_wrapper&& other) : _impl{std::move(other._impl)} {}
+
+    function_wrapper(const function_wrapper&) = delete;
+
+    function_wrapper(function_wrapper&) = delete;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:89]: Looks like copy constructor is done twice and move assignment was missed

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        while (!_task_queue.try_pop(task) && !_done) { _cv.wait(l); }
+      }
+      if (_done) { break; }
+      task();
+    }
+  }
+
+  std::atomic_bool _done;
+  std::condition_variable _cv;
+  mutable std::mutex _mut;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:134]: I dont think this needs to be mutable

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  };
+
+  void worker()
+  {
+    while (!_done)
+    {
+      function_wrapper task;
+      {
+        std::unique_lock<std::mutex> l(_mut);
+        while (!_task_queue.try_pop(task) && !_done) { _cv.wait(l); }
```
Comment (zwd-ms) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:125]: can we move the cv to `thread_safe_queue` and use a `wait_and_pop` instead?
I think that'll make the code logic cleaner

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 1 above)
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:23]: that one is a pointer queue though

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 3 above)
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:169]: maybe you don't want to branch your code based on zero threads vs more than one thread

plus this ensures that you don't just hang if you initialize with zero by mistake

Comment (jackgerrits) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:169]: In my opinion initializing with zero threads should be an error. If the user doesn't want threading they shouldnt use a thread pool. It's not up to the thread pool to implement that

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:23]: Pointers are movable so usage of that type can probably be replaced to one common implementation that satisfies both

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 3 above)
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:169]: I like to think as scaling naturally with the available resources without having to have a different code path (similar to std async)

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 1 above)
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:23]: who owns the objects though?

Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:23]: might need to change its name from ptr_queue

File: vowpalwabbit/core/include/vw/core/thread_pool.h [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+  ~thread_pool()
+  {
+    _done = true;
+    _cv.notify_all();
+  }
+
+  // Enqueues a func task in the work queue for worker threads to execute.
+  template <typename F, typename... Args>
+  std::future<typename std::result_of<F(Args...)>::type> submit(F&& func, Args&&... args)
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/thread_pool.h:163]: `std::result_of` was deprecated in 17 and removed in 20 so we should not use it.

We can instead use a trailing return type and decltype of funcs invocation. See here: https://github.com/mtrebi/thread-pool/blob/master/include/ThreadPool.h#L79

File: test/unit_test/thread_pool_test.cc [hunk 9]
```diff
@@ -0,0 +1,115 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/thread_pool.h"
```
Comment (jackgerrits) [inline test/unit_test/thread_pool_test.cc:5]: Please add tests to core unit tests and not the boost project

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 6 above)
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:134]: that's true it isn't used in any read only functions

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:23]: I dont think pointers need special treatment. We should be able to reuse the same implementation

File: vowpalwabbit/core/include/vw/core/thread_pool.h (hunk 7 above)
Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:125]: We can see what it looks like after I try out @jackgerrits suggestion to merge the existing queue's but I'll probably make it wait on `_task_queue.empty()` in that case. This way it is a good separation of concerns, the queue should not care when the thread pool is done

Comment (olgavrou) [reply vowpalwabbit/core/include/vw/core/thread_pool.h:125]: Hm that queue has set_done actually now that i think about it so might be doable

**Review by jackgerrits — APPROVED** [review]

**Review by zwd-ms — APPROVED** [review]

**Pull request merged**